File: src/main/java/com/ggemo/bilidanmakuclient/http/response/AbstractResponse.java
package com.ggemo.bilidanmakuclient.http.response;

import com.ggemo.bilidanmakuclient.http.exception.BiliClientException;
import lombok.AllArgsConstructor;
import lombok.NoArgsConstructor;
import lombok.Setter;

@Setter
@NoArgsConstructor
@AllArgsConstructor
public class AbstractResponse<T extends ResponseData> implements Response {
    int code;
    String message;
    T data;

    public static Response parse(String json) {
        throw new UnsupportedOperationException();
    }

    public int getCode() {
        return code;
    }

    public String getMessage() {
        return message;
    }

    public T getData() throws BiliClientException {
        if (this.code != 0) {
            throw new BiliClientException("code != 0");
        }
        if (data == null) {
            throw new BiliClientException("data is null");
        }
        return data;
    }

    @Override
    public String toString() {
        if (data != null) {
            return "AbstractResponse{" +
                    "code=" + code +
                    ", message='" + message + '\'' +
                    ", data=" + data +
                    '}';
        }else{
            return "AbstractResponse{" +
                    "code=" + code +
                    ", message='" + message + '\'' +
                    ", data=null" +
                    '}';
        }
    }
}

File: src/main/java/com/ggemo/va/bilidanmakuclient/http/response/AbstractResponse.kt
package com.ggemo.va.bilidanmakuclient.http.response

import com.alibaba.fastjson.annotation.JSONField
import com.ggemo.va.bilidanmakuclient.http.exception.BiliClientException

open class AbstractResponse<T : ResponseData?> : Response {
    var code: Number? = null
    @JSONField(name = "msg")
    var message: String? = null
    var data: T? = null

    @Throws(BiliClientException::class)
    fun getD(): T {
        if (this.code != 0) {
            throw BiliClientException("code != 0")
        }
        if (data == null) {
            throw BiliClientException("data is null")
        }
        return data!!
    }

    override fun toString(): String {
        return if (data != null) {
            "AbstractResponse{" +
                    "code=" + code +
                    ", message='" + message + '\'' +
                    ", data=" + data +
                    '}'
        } else {
            "AbstractResponse{" +
                    "code=" + code +
                    ", message='" + message + '\'' +
                    ", data=null" +
                    '}'
        }
    }

    companion object {
        fun parse(json: String?): Response {
            throw UnsupportedOperationException()
        }
    }
}